File: everything_parsing/27-ll1-error-recovery/js/recoveringparser.test.mjs
// node js/recoveringparser.test.mjs

import { that, equals, summary } from "./assert.mjs";
import { Grammar } from "./grammar.mjs";
import { RecoveringParser } from "./recoveringparser.mjs";

console.log("recoveringparser.test.mjs");

const toks = s => s.split(" ");

const g = Grammar.parse(`
%start E
E  -> T Ep
Ep -> + T Ep | epsilon
T  -> F Tp
Tp -> * F Tp | epsilon
F  -> ( E ) | id | num
`);
const p = new RecoveringParser(g);

const ok = p.parse(toks("id + id * id"));
that(ok.clean(), "id + id * id: accepted, no errors");

const r1 = p.parse(toks("id + * id"));
that(r1.accepted, "id + * id recovers to end of input");
equals(r1.errors.length, 1, "one error");
that(r1.errors[0].message.includes("discarding '*'"), "the stray '*' is discarded");
equals(r1.errors[0].pos, 2, "error reported at position 2");

const r2 = p.parse(toks("( id + )"));
that(r2.accepted, "( id + ) recovers");
that(r2.errors[0].message.includes("skipping T"), "T is skipped");

const r3 = p.parse(toks("( id + id"));
that(r3.accepted, "( id + id recovers");
that(r3.errors[0].message.includes("inserted missing ')'"), "the ')' is inserted");

const r4 = p.parse(toks("+ id"));
that(r4.accepted, "+ id recovers");
equals(r4.errors[0].pos, 0, "error at position 0");

const r5 = p.parse(toks("id id + id"));
that(r5.accepted, "id id + id recovers");
equals(r5.errors.length, 1, "one error");

const r6 = p.parse(toks("* * * *"));
that(r6 != null, "pathological input still returns");
that(r6.errors.length > 0, "and reports errors");

that(p.syncSet("T").has(")"), "FOLLOW(T) contains )");
that(p.syncSet("T").has("+"), "FOLLOW(T) contains +");

summary();

File: everything_parsing/27-ll1-error-recovery/java/RecoveringParserTest.java
import java.util.Arrays;
import java.util.List;

/** java -cp java/out RecoveringParserTest */
public final class RecoveringParserTest {

    static List<String> toks(String s) { return Arrays.asList(s.split(" ")); }

    public static void main(String[] args) {
        System.out.println("RecoveringParserTest");

        Grammar g = Grammar.parse("""
            %start E
            E  -> T Ep
            Ep -> + T Ep | epsilon
            T  -> F Tp
            Tp -> * F Tp | epsilon
            F  -> ( E ) | id | num
            """);
        RecoveringParser p = new RecoveringParser(g);

        // ── a clean parse has zero errors ──
        RecoveringParser.Result ok = p.parse(toks("id + id * id"));
        Assert.that(ok.clean(), "id + id * id: accepted, no errors");

        // ── panic mode: discard a stray token ──
        RecoveringParser.Result r1 = p.parse(toks("id + * id"));
        Assert.that(r1.accepted(), "id + * id recovers to end of input");
        Assert.equals(r1.errors().size(), 1, "one error");
        Assert.that(r1.errors().get(0).message().contains("discarding '*'"), "the stray '*' is discarded");
        Assert.equals(r1.errors().get(0).pos(), 2, "error reported at position 2");

        // ── panic mode: sync on FOLLOW ──
        RecoveringParser.Result r2 = p.parse(toks("( id + )"));
        Assert.that(r2.accepted(), "( id + ) recovers");
        Assert.that(r2.errors().get(0).message().contains("skipping T"), "T is skipped (')' in FOLLOW(T))");

        // ── phrase-level: insert a missing terminal ──
        RecoveringParser.Result r3 = p.parse(toks("( id + id"));
        Assert.that(r3.accepted(), "( id + id recovers");
        Assert.that(r3.errors().get(0).message().contains("inserted missing ')'"), "the ')' is inserted");

        // ── leading operator ──
        RecoveringParser.Result r4 = p.parse(toks("+ id"));
        Assert.that(r4.accepted(), "+ id recovers");
        Assert.equals(r4.errors().get(0).pos(), 0, "error at position 0");

        // ── extra operand (missing operator) ──
        RecoveringParser.Result r5 = p.parse(toks("id id + id"));
        Assert.that(r5.accepted(), "id id + id recovers");
        Assert.equals(r5.errors().size(), 1, "one error: the second id has no operator before it");

        // ── recovery always terminates (guard never trips for these) ──
        RecoveringParser.Result r6 = p.parse(toks("* * * *"));
        Assert.that(r6 != null, "pathological input still returns");
        Assert.that(!r6.errors().isEmpty(), "and reports errors");

        // ── sync sets are FOLLOW sets ──
        Assert.that(p.syncSet("T").contains(")"), "FOLLOW(T) contains )");
        Assert.that(p.syncSet("T").contains("+"), "FOLLOW(T) contains +");

        Assert.summary();
    }
}
